### PR TITLE
Extend RelayCall to inspect the call response frame

### PR DESCRIPTION
**Pull request by cinchurge**
In order to support features where inspection of the response frame is required, extend `RelayCall` with an additional `CallResponse()` method which can be used for inspecting the response frame of a call. A new interface `RespFrame` is added which mirrors the `CallFrame` passed into `RelayHost.Start()`.

### CODE REVIEW

File: relay.go [hunk 1]
```diff
@@ -552,6 +552,15 @@ func (r *Relayer) handleNonCallReq(f *Frame) error {
 		return nil
 	}
 
+	if f.messageType() == messageTypeCallRes {
+		cr, err := newLazyCallRes(f)
+		if err != nil {
+			r.logger.Error(err.Error())
```
Comment (prashantv) [inline relay.go:558]: This won't have much useful context if we log errors in this way, can we have more information:
 * a message indicating that a call res frame was not parsed
 * context on the call (e.g., the ID, any info about the item that may be useful)

File: relay.go [hunk 2]
```diff
@@ -552,6 +552,15 @@ func (r *Relayer) handleNonCallReq(f *Frame) error {
 		return nil
 	}
 
+	if f.messageType() == messageTypeCallRes {
```
Comment (prashantv) [inline relay.go:555]: since we hae another `messageType()` check below, can we use a switch for both these cases?

File: relay/relay.go [hunk 3]
```diff
@@ -67,6 +67,18 @@ type CallFrame interface {
 	Arg2Append(key, val []byte)
 }
 
+// RespFrame is an interface that abstracts access to the CallRes frame
+type RespFrame interface {
+	// OK indicates whether the call was successful
+	OK() bool
+
+	// Arg2 returns the raw arg2 payload
+	Arg2() []byte
```
Comment (prashantv) [inline relay/relay.go:76]: Do we want to expose arg2?

File: relay_api.go [hunk 4]
```diff
@@ -20,7 +20,9 @@
 
 package tchannel
 
-import "github.com/uber/tchannel-go/relay"
+import (
+	"github.com/uber/tchannel-go/relay"
+)
```
Comment (prashantv) [inline relay_api.go:25]: nit: can revert this change

File: relay_messages.go [hunk 5]
```diff
@@ -272,3 +342,13 @@ func finishesCall(f *Frame) bool {
 		return false
 	}
 }
+
+// isCallResOK indicates whether the call was successful
+func isCallResOK(f *Frame) bool {
+	return f.Payload[_resCodeIndex] == _resCodeOK
+}
```
Comment (prashantv) [inline relay_messages.go:349]: should we use this method to implement `lazyCallRes.OK` (so we don't have 2 different ways to read this field)

also means we can drop a few fields from the `lazyCallRes` struct

File: relay_messages_test.go [hunk 6]
```diff
@@ -161,58 +162,73 @@ func withLazyCallReqCombinations(f func(cr testCallReq)) {
...
-		params.headers = map[string]string{
-			"k1":      "v1",
-			"k222222": "",
-			"k3":      "thisisalonglongkey",
-		}
+		params.headers["k1"] = "v1"
+		params.headers["k222222"] = ""
+		params.headers["k3"] = "thisisalonglongkey"
+	}
+	if cr&(resHasArg2|resHasFragmentedArg2) != 0 {
```
Comment (prashantv) [inline relay_messages_test.go:206]: why do we check `HasFragmentedArg2` here?

File: relay_test.go [hunk 7]
```diff
@@ -1420,6 +1420,33 @@ func TestRelayTransferredBytes(t *testing.T) {
 	})
 }
 
+func TestRelayCallResponse(t *testing.T) {
```
Comment (prashantv) [inline relay_test.go:1423]: Can we add an integration test that sends response headers, and validates them via the `RespFrameFn`?

File: relay/relay.go (hunk 3 above)
Comment (cinchurge) [reply relay/relay.go:76]: this is primarily for testing since the `thrift` package imports `tchannel`, and we can't really test arg2 iteration with the raw scheme. an alternative is to do a type assertion and inspect `lazyCallRes` directly, but even that would require exporting arg2 somehow. is there any other way we can test without exporting an arg2 accessor?

File: relay_api.go (hunk 4 above)
Comment (cinchurge) [reply relay_api.go:25]: not sure why? we're passing in `relay.RespFrame` below

File: relay_messages_test.go (hunk 6 above)
Comment (cinchurge) [reply relay_messages_test.go:206]: so the arg scheme is set to `thrift`. we only support arg2 iteration for `thrfit`, and the two types here are meant for adding key/vals to arg2

File: relay_test.go (hunk 7 above)
Comment (cinchurge) [reply relay_test.go:1423]: not sure if i'm missing something - in this test the response header is set by the echo handler, copied over from the call, and we verify that the bytes in arg2 are exactly what we send. is there something else you had in mind?

File: relay.go [hunk 8]
```diff
@@ -552,19 +552,24 @@ func (r *Relayer) handleNonCallReq(f *Frame) error {
 		return nil
 	}
 
-	if f.messageType() == messageTypeCallRes {
+	switch f.messageType() {
+	case messageTypeCallRes:
+		// Invoke call.CallResponse() if we got the response frame
```
Comment (prashantv) [inline relay.go:557]:
```suggestion
		// Invoke call.CallResponse() if we get a valid call response frame.
```

File: relay.go [hunk 9]
```diff
@@ -552,19 +552,24 @@ func (r *Relayer) handleNonCallReq(f *Frame) error {
...
 		cr, err := newLazyCallRes(f)
-		if err != nil {
-			r.logger.Error(err.Error())
-		} else {
+		if err == nil {
 			item.call.CallResponse(cr)
+		} else {
+			r.logger.WithFields(
```
Comment (prashantv) [inline relay.go:562]: can we get coverage for this?

File: relay/relay.go (hunk 3 above)
Comment (prashantv) [reply relay/relay.go:76]: ahh, I see. What are we doing for testing arg2 on the call req, can we do the same?

One potential follow up to maybe simplify things, we could change both the lazyCallReq and lazyCallRes to only export `Arg2() []byte` and `ArgScheme() []byte`, and then have the caller check the scheme, and if it's thrift, they call `NewKeyValueIterator` on the `Arg2() []byte` directly,
https://pkg.go.dev/github.com/uber/tchannel-go@v1.21.2/thrift/arg2#NewKeyValIterator

File: relay_api.go (hunk 4 above)
Comment (prashantv) [reply relay_api.go:25]: Sorry, what I meant is that the import matches the same as what it was previously, but it's grouped (which isn't necessary since there's only one import)

File: relay_messages_test.go (hunk 6 above)
Comment (prashantv) [reply relay_messages_test.go:206]: ah, would it be clearer to use a separate orthogonal bit like `resIsThrift` since it is possible to have a fragmented arg2 without thrift (though we may not necessarily find it useful in our tests)

File: relay_test.go (hunk 7 above)
Comment (prashantv) [reply relay_test.go:1423]: What I was thinking was that instead of verifying the arg2 bytes, we actually response with thift response headers so arg2 is an encoded header map, and we validate the values map.

I think we do something like that on the request side:
https://github.com/uber/tchannel-go/blob/dev/relay_test.go#L1233

File: relay/relay.go (hunk 3 above)
Comment (cinchurge) [reply relay/relay.go:76]: we do export an accessor for the arg2 offset in the callReq, but it really doesn't feel any better than just exporting the slice directly. I do think though that exposing `Arg2()` and `Scheme()` is a good idea. I'll add a TODO for that.

Comment (cinchurge) [reply relay/relay.go:79]: Added `Scheme()` to `RespFrame`

File: relay_messages.go [hunk 10]
```diff
@@ -80,17 +79,81 @@ func (e lazyError) Code() SystemErrCode {
...
+
+	rbuf := typed.NewReadBuffer(f.SizedPayload())
+	rbuf.SkipBytes(1)           // flags
+	rbuf.SkipBytes(1)           // code
+	rbuf.SkipBytes(_spanLength) // tracing
+
+	var as []byte
+	nh := int(rbuf.ReadSingleByte())
+	for i := 0; i < nh; i++ {
+		hk := int(rbuf.ReadSingleByte())
```
Comment (prashantv) [inline relay_messages.go:101]: not sure what `hk` stands for, perhaps `keyLen` and `valLen`?

File: relay_test.go [hunk 11]
```diff
@@ -34,6 +34,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/uber/tchannel-go/thrift/arg2"
```
Comment (prashantv) [inline relay_test.go:37]: nit: import ordering

File: relay_test.go [hunk 12]
```diff
@@ -931,6 +933,68 @@ func TestRelayStalledClientConnection(t *testing.T) {
...
+		defer client.Close()
+
+		ctx, cancel := NewContext(testutils.Timeout(time.Second))
+		defer cancel()
+
+		// Data to fit one frame fully, but large enough that a number of these frames will fill
+		// all the buffers and cause the relay to drop the response frame. Buffers are:
+		// 1. Relay's sendCh on the connection to the client (set to 10 frames explicitly)
+		// 2. Relay's TCP send buffer for the connection to the client.
+		// 3. Client's TCP receive buffer on the connection to the relay.
```
Comment (prashantv) [inline relay_test.go:981]: don't think this is relevant to this test? we don't need any drops to happen for this test right?

**Review by prashantv — APPROVED** [review]

File: relay_test.go (hunk 12 above)
Comment (cinchurge) [reply relay_test.go]: sorry, copypasta
